### PR TITLE
wallet2: export_outputs + import_outputs + cold signing patches [release]

### PR DESCRIPTION
**Pull request by j-berman**
PR of the code discussed in #8484 (the base patches in this PR are written by @moneromooo-monero). This PR should also patch #8476:

- Only offline wallets can import outputs.
- Empty exports won't corrupt wallet files when signing.
- Subaddress indexes are exported & imported.

I added explicit tests for the buggy failure cases described in #8476. But I wasn't able to reproduce @woodser's issue described in [this comment](https://github.com/monero-project/monero/issues/8484#issuecomment-1213632430):

> it occasionally hangs on import_outputs (when used correctly with view-only and offline wallets) for some unknown reason

Requesting @woodser check if he can repro the hang with this code; I suspect the hang might have something to do with the slightly off empty exports, which should be good with this PR.

### CODE REVIEW

File: src/wallet/wallet2.h [hunk 1]
```diff
@@ -401,16 +401,20 @@ namespace tools
...
-        VERSION_FIELD(0)
+        VERSION_FIELD(1)
         FIELD(m_pubkey)
         VARINT_FIELD(m_internal_output_index)
         VARINT_FIELD(m_global_output_index)
         FIELD(m_tx_pubkey)
         FIELD(m_flags.flags)
         VARINT_FIELD(m_amount)
         FIELD(m_additional_tx_keys)
+        VARINT_FIELD(m_subaddr_index_major)
```
Comment (moneromooo-monero) [inline src/wallet/wallet2.h:416]: In theory I should have added a if (version < 1) return false; (since you can't really "default" to a sane value).

File: tests/functional_tests/functional_tests_rpc.py [hunk 2]
```diff
@@ -82,6 +82,12 @@
   outputs.append(open(FUNCTIONAL_TESTS_DIRECTORY + '/wallet' + str(i) + '.log', 'a+'))
   ports.append(18090+i)
 
+# 1 offline wallet
+command_lines.append([str(18090+N_WALLETS) if x == "wallet_port" else x for x in wallet_base])
+command_lines[-1] += ["--offline"]
+outputs.append(open(FUNCTIONAL_TESTS_DIRECTORY + '/wallet' + str(N_WALLETS) + '.log', 'a+'))
+ports.append(18090+N_WALLETS)
+
```
Comment (moneromooo-monero) [inline tests/functional_tests/functional_tests_rpc.py:90]: Any chance this cam be moved back to the loop since it's very similar ? It's just a if i == N_WALLETS-1:  command_lines[i] += "--offline"; AFAICT.

**Review by moneromooo-monero — COMMENTED** [review]
Thanks for doing that.

**Review by moneromooo-monero — APPROVED** [review]

**Comment by woodser** [comment]
With this PR I'm getting `Hot wallets cannot import outputs`, even though the wallet is offline and initialized with an empty "" server. I guess `m_offline` is false when it should be true without the server?

Side note, the error message is `Failed to import outputsHot wallets cannot import outputs` because there's no `: ` on [this line](https://github.com/monero-project/monero/blob/b6a029f222abada36c7bc6c65899a4ac969d7dee/src/wallet/wallet2.cpp#L13398).

**Comment by j-berman** [comment]
Thinking more on this, it seems like I'm missing more ways that a wallet can technically be offline even if `--offline` isn't passed at startup or if `set_offline` isn't explicitly called (e.g. wi-fi turned off). An empty "" server seems like one (imo somewhat ambiguous) way of multiple to start a cold wallet.

I'm thinking the best option on the table is to require cold wallet implementations to explicitly set the offline flag (either when starting the wallet, or by calling `set_offline`/`setOffline` via the wallet API). And I'm thinking the error message could be more detailed for implementers:

```
Hot wallets cannot import outputs (you must pass the --offline flag when starting the wallet or explicitly set the wallet to the offline mode)
```

Maybe @moneromooo-monero has a different opinion on this.

(I also cleaned up the spacing; good spot, thank you)

**Comment by j-berman** [comment]
Actually, perhaps there should even be a new required flag `--cold-wallet` which permanently ensures `m_offline` is false after the wallet is created. Otherwise it would be possible to trigger [this error](https://github.com/moneromooo-monero/bitmonero/commit/79e09fbf8c5f04a7f308762a5bc7a0927a4cd909#r80699311) (and other unexpected behavior) like this:

1. Create a full wallet
2. Set `m_offline` to true
3. Import outputs
4. Set `m_offline` to false
5. Transfer

**Comment by moneromooo-monero** [comment]
Do you actually want to import outputs into a hot wallet ? If so, why ? If not, can you rephrase ?

**Comment by woodser** [comment]
> Do you actually want to import outputs into a hot wallet ? If so, why ? If not, can you rephrase ?

I don't have any use case for that, but the error I'm seeing `Hot wallets cannot import outputs` isn't correct since the wallet is offline, not connected to any server.

**Comment by moneromooo-monero** [comment]
OK, then I guess it can be made... "Only cold wallets can import outputs". Since it's what tests for IIRC.

**Comment by woodser** [comment]
> OK, then I guess it can be made... "Only cold wallets can import outputs". Since it's what tests for IIRC.

So with this change, a user would need to manually call `set_offline` in addition to passing an empty server, to denote the wallet as permanently cold?

**Comment by moneromooo-monero** [comment]
Looking at the code, I see it's using m_offline, migt have been because there is no easy way to tell cold wallet then... I thouht it was testing for cold, but there doesn't seem to be a way. In that case, the best might be to add a "make_cold" command, which would set a flag in wallet2, which could hten be tested. And the wallet could defualt to offline when this flag is set.
Do you see a drawback here ?

**Comment by j-berman** [comment]
Agree, I don't see a better option than an explicit cold wallet flag

The primary drawback I can see is that cold wallet setups and implementations will break (they will need to add a flag to start a cold wallet). I would think wallets that implement cold signing should probably be notified and we'll need to e.g. update documentation [like this](https://monerodocs.org/cold-storage/offline-transaction-signing/) too

**Comment by moneromooo-monero** [comment]
We could keep track of "have I ever refreshed blocks from a daemon". If so, reject attempt to import outputs.

**Comment by woodser** [comment]
> We could keep track of "have I ever refreshed blocks from a daemon". If so, reject attempt to import outputs.

I like something like this in the spirit of simplicity and backwards compatibility. If a client initializes a wallet without the server uri, it's offline, and I would think it's eligible to import outputs. Otherwise, it'll be an extra step and breaking change for anyone currently using offline wallets.

**Comment by j-berman** [comment]
+1^ sounds like a better solution

**Comment by moneromooo-monero** [comment]
I added this check in #8513

**Comment by woodser** [comment]
> I added this check in #8513

I'm getting an error "Nothing requested" when exporting outputs from a newly synced view-only wallet.

**Comment by j-berman** [comment]
Closing this in favor of #8513 to consolidate discussion
